prov/efa/src/rxr/rxr_pkt_type_rts.c

```diff
@@ -0,0 +1,783 @@
+						     tx_entry->iov, tx_entry->iov_count, 0);
+			assert(data_len == tx_entry->total_len);
+			rts_hdr->flags |= RXR_SHM_HDR_DATA;
+			pkt_entry->pkt_size += data_len;
+		} else {
+			/* rendezvous protocol
+			 * place iov_count first, then local iov
+			 */
+			memcpy(data, &tx_entry->iov_count, sizeof(size_t));
+			data += sizeof(size_t);
+			pkt_entry->pkt_size += sizeof(size_t);
+			memcpy(data, tx_entry->iov, sizeof(struct iovec) * tx_entry->iov_count);
+			pkt_entry->pkt_size += sizeof(struct iovec) * tx_entry->iov_count;
+		}
+	} else {
+		/* will be sent over efa provider */
+		data_len = ofi_copy_from_iov(data, mtu - pkt_entry->pkt_size,
+					     tx_entry->iov, tx_entry->iov_count, 0);
+		assert(data_len == rxr_get_rts_data_size(ep, rts_hdr));
+		pkt_entry->pkt_size += data_len;
+	}
+
+	assert(pkt_entry->pkt_size <= mtu);
+	pkt_entry->addr = tx_entry->addr;
+	pkt_entry->x_entry = (void *)tx_entry;
+	return 0;
+}
+
+/*
+ *  rxr_pkt_handle_rts_sent() and related functions
+ */
+void rxr_pkt_handle_rts_sent(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_peer *peer;
+	struct rxr_tx_entry *tx_entry;
+	size_t data_sent;
+
+	tx_entry = (struct rxr_tx_entry *)pkt_entry->x_entry;
+
+	peer = rxr_ep_get_peer(ep, pkt_entry->addr);
+	assert(peer);
+	if (tx_entry->op == ofi_op_read_req) {
+		tx_entry->bytes_sent = 0;
+		tx_entry->state = RXR_TX_WAIT_READ_FINISH;
+		return;
+	}
+
+	data_sent = rxr_get_rts_data_size(ep, rxr_get_rts_hdr(pkt_entry->pkt));
+
+	tx_entry->bytes_sent += data_sent;
+
+	if ((rxr_env.enable_shm_transfer && peer->is_local) ||
+	    !(efa_mr_cache_enable && tx_entry->total_len > data_sent))
+		return;
+
+	/*
+	 * Register the data buffers inline only if the application did not
+	 * provide a descriptor with the tx op
+	 */
+	if (rxr_ep_mr_local(ep) && !tx_entry->desc[0])
+		rxr_inline_mr_reg(rxr_ep_domain(ep), tx_entry);
+}
+
+/*
+ *  rxr_pkt_handle_rts_recv() and related functions
+ */
+char *rxr_pkt_read_rts_base_hdr(struct rxr_ep *ep,
+				struct rxr_rx_entry *rx_entry,
+				struct rxr_pkt_entry *pkt_entry)
+{
+	char *data;
+	struct rxr_rts_hdr *rts_hdr = NULL;
+	/*
+	 * Use the correct header and grab CQ data and data, but ignore the
+	 * source_address since that has been fetched and processed already
+	 */
+
+	rts_hdr = rxr_get_rts_hdr(pkt_entry->pkt);
+
+	rx_entry->addr = pkt_entry->addr;
+	rx_entry->tx_id = rts_hdr->tx_id;
+	rx_entry->msg_id = rts_hdr->msg_id;
+	rx_entry->total_len = rts_hdr->data_len;
+	rx_entry->cq_entry.tag = rts_hdr->tag;
+
+	if (rts_hdr->flags & RXR_REMOTE_CQ_DATA) {
+		rx_entry->cq_entry.flags |= FI_REMOTE_CQ_DATA;
+		data = rxr_get_ctrl_cq_pkt(rts_hdr)->data + rts_hdr->addrlen;
+		rx_entry->cq_entry.data =
+				rxr_get_ctrl_cq_pkt(rts_hdr)->hdr.cq_data;
+	} else {
+		rx_entry->cq_entry.data = 0;
+		data = rxr_get_ctrl_pkt(rts_hdr)->data + rts_hdr->addrlen;
+	}
+
+	return data;
+}
+
+static
+char *rxr_pkt_read_rts_rma_hdr(struct rxr_ep *ep,
```

Comment:
I'd switch this from 'read' to 'proc' as well.

Reply:
Done.